### PR TITLE
[REVIEW] switch tests/benchmarks to use `pool_memory_resource`

### PR DESCRIPTION
**Pull request by rongou**
@harrism @kkraus14

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/benchmarks/fixture/benchmark_fixture.hpp [hunk 1]
```diff
@@ -54,8 +55,10 @@ class benchmark : public ::benchmark::Fixture {
  public:
   virtual void SetUp(const ::benchmark::State& state)
   {
-    auto mr = new rmm::mr::cnmem_memory_resource;
-    rmm::mr::set_default_resource(mr);  // set default resource to cnmem
+    using cuda_mr = rmm::mr::cuda_memory_resource;
+    using pool_mr = rmm::mr::pool_memory_resource<cuda_mr>;
+    auto mr       = new pool_mr(new cuda_mr());
```
Comment (jrhemstad) [inline cpp/benchmarks/fixture/benchmark_fixture.hpp:60]: The `cuda_mr` will leak here.

File: cpp/tests/utilities/base_fixture.hpp [hunk 2]
```diff
@@ -237,9 +237,12 @@ class TempDirTestEnvironment : public ::testing::Environment {
...
   std::string const &allocation_mode)
 {
-  if (allocation_mode == "cuda") return std::make_unique<rmm::mr::cuda_memory_resource>();
-  if (allocation_mode == "pool") return std::make_unique<rmm::mr::cnmem_memory_resource>();
-  if (allocation_mode == "managed") return std::make_unique<rmm::mr::managed_memory_resource>();
+  using cuda_mr    = rmm::mr::cuda_memory_resource;
+  using pool_mr    = rmm::mr::pool_memory_resource<cuda_mr>;
+  using managed_mr = rmm::mr::managed_memory_resource;
+  if (allocation_mode == "cuda") return std::make_unique<cuda_mr>();
+  if (allocation_mode == "pool") return std::make_unique<pool_mr>(new cuda_mr());
```
Comment (jrhemstad) [inline cpp/tests/utilities/base_fixture.hpp:244]: Same here. `cuda_mr` will leak.

**Comment by jrhemstad** [comment]
add to whitelist

**Comment by rongou** [comment]
Hmm, seems like the best way to prevent leak is to make the `Upstream` raw pointer into either a `std::unique_ptr` or a `std::shared_ptr`. What do you think?

**Comment by jrhemstad** [comment]
> Hmm, seems like the best way to prevent leak is to make the `Upstream` raw pointer into either a `std::unique_ptr` or a `std::shared_ptr`. What do you think?

You mean modify the resources? No, that's not what I meant. We don't want to require a resource to take ownership of it's upstream.

I'd add a `unique_ptr` to the fixture class to maintain the lifetime of the upstream resource.

**Comment by rongou** [comment]
I backed out the changes to tests and benchmarks pending better lifetime management in RMM. Please take another look.

**Comment by kkraus14** [comment]
We haven't outright dropped cnmem from rmm, right? If so we still need a `librmm` shared library and to link `libcudf` to it until we do so. What are the java build issues?

**Comment by rongou** [comment]
In the JNI code we switched from CNMeM to the new `pool_memory_resource`. When I build `libcudfjni.so` on my desktop (Ubuntu 18.04), it no longer depends on `librmm.so`, but on our CI build that uses Centos7, the old linker still adds the dependency, causing problems.

I guess the quickest solution for now is to switch to `pool_memory_resource` in the tests and benchmarks, and manage the cuda mr lifetime in the fixture.

**Comment by trevorsm7** [comment]
Should this be removed as well? (minus `message(STATUS "RMM: RMM_INCLUDE set to ${RMM_INCLUDE}")`)
https://github.com/rapidsai/cudf/blob/489c3a781bb46b197f4c4cdca16c64bd55dff286/cpp/CMakeLists.txt#L280-L289

**Comment by harrism** [comment]
First step is to move cuDF to using pool_memory_resource instead of cnmem_memory_resource. I was going to open a PR to test doing this for gtests first (tried to get to it today, will try again tomorrow). Then we have to change the cuDF Python.

**Comment by harrism** [comment]
@kkraus14 I think once 5705 is merged libcudf does not need to link librmm. librmm will still be dynamically loaded when the RMM python module is imported.  Once we get corresponding PRs merged into libcuspatial, libcugraph, and libcuml we can deprecate cnmem in RMM, and hopefully remove it in 0.16.

**Comment by harrism** [comment]
> Should this be removed as well? (minus `message(STATUS "RMM: RMM_INCLUDE set to ${RMM_INCLUDE}")`)

We still need RMM_INCLUDE, we just don't need to link a library.

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -386,8 +386,7 @@ endif(CONDA_INCLUDE_DIRS)
 link_directories("${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES}" # CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES is an undocumented/unsupported variable containing the link directories for nvcc
                  "${CMAKE_BINARY_DIR}/lib"
                  "${CMAKE_BINARY_DIR}"
-                 "${GTEST_LIBRARY_DIR}"
-                 "${RMM_LIBRARY}")
```
Comment (kkraus14) [inline cpp/CMakeLists.txt:390]: If we're removing this, can we remove the corresponding `find_library` call for rmm as well?

**Comment by trevorsm7** [comment]
> > Should this be removed as well? (minus `message(STATUS "RMM: RMM_INCLUDE set to ${RMM_INCLUDE}")`)
> 
> We still need RMM_INCLUDE, we just don't need to link a library.

> If we're removing this, can we remove the corresponding find_library call for rmm as well?

That's what I was asking about in the lines I selected above. The add_library call could be removed as well then

File: cpp/CMakeLists.txt (hunk 3 above)
Comment (rongou) [reply cpp/CMakeLists.txt:390]: Done.

File: cpp/benchmarks/fixture/benchmark_fixture.hpp (hunk 1 above)
Comment (rongou) [reply cpp/benchmarks/fixture/benchmark_fixture.hpp]: Fixed.

File: cpp/tests/utilities/base_fixture.hpp (hunk 2 above)
Comment (rongou) [reply cpp/tests/utilities/base_fixture.hpp]: Added temporary fix.

**Comment by harrism** [comment]
@rongou looks like you changed the purpose and title of this PR, and now it duplicates changes in #5705?

#5705 is waiting on rapidsai/rmm#444

**Comment by rongou** [comment]
This was the original title. My original goal was to fix the java CI that doesn't like linking to `librmm.so` even though we no longer use any symbols from there. It's fine to close this as long as it gets fixed.

**Comment by rongou** [comment]
rerun tests

**Comment by harrism** [comment]
Replaced by #5864

**Pull request closed without merging**